### PR TITLE
[spark] Support nested col pruning

### PR DESCRIPTION
**Pull request by Zouxxyy**
### Purpose

to #4209, Support nested col pruning,  e.g.

```sql
CREATE TABLE students (
    name STRING,
    age INT,
    course STRUCT<course_name: STRING, grade: DOUBLE>
) USING paimon;
```

```sql
SELECT course.grade FROM students;
```

will only obtain `course.grade` from colume-storage-format (parquet, orc)

### CODE REVIEW

**Review by JingsongLi — COMMENTED** [review]
Are there some configuration to disable nested projection? I am concern about bugs in nested projection, at least, we should have option to disable it.

**Comment by Zouxxyy** [comment]
> Are there some configuration to disable nested projection? I am concern about bugs in nested projection, at least, we should have option to disable it.

yes, spark has a conf ti disable it

```scala
  val NESTED_SCHEMA_PRUNING_ENABLED =
    buildConf("spark.sql.optimizer.nestedSchemaPruning.enabled")
      .internal()
      .doc("Prune nested fields from a logical relation's output which are unnecessary in " +
        "satisfying a query. This optimization allows columnar file format readers to avoid " +
        "reading unnecessary nested column data. Currently Parquet and ORC are the " +
        "data sources that implement this optimization.")
      .version("2.4.1")
      .booleanConf
      .createWithDefault(true)
```

File: paimon-format/src/main/java/org/apache/paimon/format/parquet/ParquetReaderFactory.java [hunk 1]
```diff
@@ -155,13 +156,37 @@ private MessageType clipParquetSchema(GroupType parquetSchema) {
...
 
+    /** Clips `parquetGroup` by `requestedRowType`. */
+    private GroupType clipParquetGroup(GroupType parquetGroup, RowType requestedRowType) {
+        Types.GroupBuilder<GroupType> builder = Types.buildGroup(parquetGroup.getRepetition());
+        List<DataField> fields = requestedRowType.getFields();
+        for (DataField field : fields) {
+            String fieldName = field.name();
+            if (parquetGroup.containsField(fieldName)) {
+                Type type = parquetGroup.getType(fieldName);
+                if (type instanceof GroupType && field.type() instanceof RowType) {
```
Comment (JingsongLi) [inline paimon-format/src/main/java/org/apache/paimon/format/parquet/ParquetReaderFactory.java:178]: Can Spark push down nested fields for array and map?

Comment (Zouxxyy) [reply paimon-format/src/main/java/org/apache/paimon/format/parquet/ParquetReaderFactory.java:178]: Not support, only support  struct col pruning

image

image

Comment (Zouxxyy) [reply paimon-format/src/main/java/org/apache/paimon/format/parquet/ParquetReaderFactory.java]: It support, update

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**